### PR TITLE
Fixed BCE Loss

### PR DESCRIPTION
**Pull request by abh2k**
Fixec BCE Loss.

### CODE REVIEW

**Comment by abh2k** [comment]
The cross_entropy function was actually the BCE Loss  function. I  added the reduction step also.

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss.hpp [hunk 1]
```diff
@@ -75,6 +74,10 @@ class CrossEntropyError
   double Eps() const { return eps; }
   //! Modify the epsilon.
   double& Eps() { return eps; }
+  //! Get the reduction.
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss.hpp:77]: Can you add an empty line between 76 and 77.

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss.hpp [hunk 2]
```diff
@@ -88,12 +91,13 @@ class CrossEntropyError
 
   //! The minimum value used for computing logarithms and denominators
   double eps;
-}; // class CrossEntropyError
+  bool reduction;
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss.hpp:94]: Parameter description is missing .

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss.hpp [hunk 3]
```diff
@@ -31,16 +30,16 @@ template <
     typename InputDataType = arma::mat,
     typename OutputDataType = arma::mat
 >
-class CrossEntropyError
+class BCELoss
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss.hpp:33]: We should provide an alias for the class to keep backward compatibility.

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss.hpp [hunk 4]
```diff
@@ -31,16 +30,16 @@ template <
...
  public:
   /**
-   * Create the CrossEntropyError object.
+   * Create the BinaryCrossEntropyLoss object.
    *
    * @param eps The minimum value used for computing logarithms
    *            and denominators in a numerically stable way.
    */
-  CrossEntropyError(const double eps = 1e-10);
+  BCELoss(const double eps = 1e-10, const bool reduction = false);
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss.hpp:42]: Parameter description is missing.

File: src/mlpack/tests/loss_functions_test.cpp [hunk 5]
```diff
@@ -268,12 +268,12 @@ TEST_CASE("SimpleMeanSquaredErrorTest", "[LossFunctionsTest]")
...
 /*
- * Simple test for the cross-entropy error performance function.
+ * Simple test for the binary-cross-entropy lossfunction.
  */
-TEST_CASE("SimpleCrossEntropyErrorTest", "[LossFunctionsTest]")
+TEST_CASE("SimpleBinaryCrossEntropyLossTest", "[LossFunctionsTest]")
 {
   arma::mat input1, input2, output, target1, target2;
-  CrossEntropyError<> module(1e-6);
+  BCELoss<> module(1e-6, false);
```
Comment (zoq) [inline src/mlpack/tests/loss_functions_test.cpp:276]: We should add a test that that uses reduction as well.

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss.hpp (hunk 3 above)
Comment (abh2k) [reply src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss.hpp:33]: Could you tell me how to do this.

**Comment by abh2k** [comment]
https://colab.research.google.com/drive/14PxGHVCzNLa9mN4VGluM2p0oQ9OCpCMt?usp=sharing
For calculation of mean i guess we have to divide the loss  with number  of elements  instead  of batch size.
am i correct?
@zoq

**Comment by abh2k** [comment]
Also  one test case fail because the default of BCELoss is kept mean.  Shall i change the default or the test case.

**Comment by abh2k** [comment]
I have made the mean to take over all elements. This  is    how  pytorch does.

**Comment by abh2k** [comment]
model = new FFN<BCELoss<>, GlorotInitialization>();
FFN<BCELoss<>, GlorotInitialization>* model;

I have a question.  I    want to pass `reduction=false` in  `BCELoss`.   How will i do  this.

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss.hpp (hunk 3 above)
Comment (zoq) [reply src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss.hpp:33]: Here is one example:

https://github.com/mlpack/mlpack/blob/56ac3d219b9dc7281ebee5b2385302943070e180/src/mlpack/methods/ann/layer/base_layer.hpp#L272-L278

here we create a alias called `GaussianFunctionLayer` which is basically a `BaseLayer` with the correct template parameter.

**Comment by zoq** [comment]
> model = new FFN<BCELoss<>, GlorotInitialization>();
> FFN<BCELoss<>, GlorotInitialization>* model;
> 
> I have a question. I want to pass `reduction=false` in `BCELoss`. How will i do this.

pass an instantiation of the loss function e.g.:

`FFN<BCELoss<>, GlorotInitialization>* model = new FFN<BCELoss<>, GlorotInitialization>(BCELoss<>(false));`

**Comment by abh2k** [comment]
@zoq  Please review. I think  I have made   all  the changes.

**Comment by abh2k** [comment]
Pleas tell  me       the error        in checks.

**Comment by zoq** [comment]
> Pleas tell me the error in checks.

Some tests are probabilistic, so it's possible that in one out of 100 runs they fail like if the threshold is to tight. I went over the failing tests and they are not related with the changes made in the PR, so all good on that front.

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss_impl.hpp [hunk 6]
```diff
@@ -13,42 +13,49 @@
...
+BCELoss<InputDataType, OutputDataType>::Forward(
     const PredictionType& prediction,
     const TargetType& target)
 {
-  return -arma::accu(target % arma::log(prediction + eps) +
-      (1. - target) % arma::log(1. - prediction + eps));
+  typedef typename PredictionType::elem_type ElemType;
+
+  ElemType loss = -arma::accu(target % arma::log(prediction + eps) + 
+      (1. - target) % arma::log(1. - prediction + eps));;
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss_impl.hpp:38]:
```suggestion
      (1. - target) % arma::log(1. - prediction + eps));
```

Remove the double `;`.

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss_impl.hpp [hunk 7]
```diff
@@ -13,42 +13,49 @@
...
+  if(reduction)
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss_impl.hpp:39]:
```suggestion
  if (reduction)
```

Minor style fix insert space after `if`.

File: src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss_impl.hpp [hunk 8]
```diff
@@ -13,42 +13,49 @@
...
 template<typename InputDataType, typename OutputDataType>
 template<typename PredictionType, typename TargetType, typename LossType>
-void CrossEntropyError<InputDataType, OutputDataType>::Backward(
+void BCELoss<InputDataType, OutputDataType>::Backward(
     const PredictionType& prediction,
     const TargetType& target,
     LossType& loss)
 {
   loss = (1. - target) / (1. - prediction + eps) - target / (prediction + eps);
+  if(reduction)
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/binary_cross_entropy_loss_impl.hpp:52]:
```suggestion
  if (reduction)
```

Minor style fix insert space after `if`.

**Review by zoq — COMMENTED** [review]
Nice, just some minor comments.

**Review by zoq — APPROVED** [review]
Thanks for putting this together, no more comments from my side.

**Comment by zoq** [comment]
Thanks for the contribution.
